feat: Item Filter GUI

ShadowCat117 opened the pull request:
Tested it a fair amount and I can't find any issues with it but I would appreciate it if someone else would give it a thorough test as I may have overlooked something.

It's a little hacky when populating the widgets with new values from the search bar and vice versa but I couldn't think of a better way to keep them in sync.

The first commit makes it so that now only screens that can use the advanced search will use it, instead they will have the original search widget.

Something I did want to add was a way to filter what filters are available on a per screen basis, for example the item guide screen has no use for the profession filter, or only the trade market can use the price filter etc.


https://github.com/Wynntils/Artemis/assets/8337467/4d699694-1fcc-4c59-aae0-cca95d3744d4

ShadowCat117 committed:
refactor: Remove InteractiveContainerType and replace with WynncraftContainer's

This also will now only support advanced searching on screens that actually use it

ShadowCat117 committed:
feat: Item Filter GUI

ShadowCat117 committed:
fix: Fix value widget disappearing after resizing window

ShadowCat117 committed:
fix: Remove extra space after filter query

ShadowCat117 commented:
@magicus  The first commit for this is a partial implementation of what you mentioned before about having different classes for each container instead of the enum, I haven't fully implemented it yet as I wanted to get your opinion on what I've done so far so for now it's just for the searchable/scrollable containers and personal storage.

Future things I would add to this approach would be a property for `FullscreenContainerProperty` for containers that replace the inventory as well, such as the ability tree and content book, then instead of checking for ability tree in the item lock feature, we won't lock the slot if the screen is a fullscreen one.
Or for crafting stations that all have the same slots for materials and ingredients, we could detect what profession it is and display a tooltip of what you will be crafting based on the items given and that could use a `CraftingStationContainer` similar to `PersonalStorageContainer`

kristofbolyai commented:
It's too late to review this today, but it is really incredible, watching the showcase video it seems everything is handled beautifully.

kristofbolyai commented on common/src/main/java/com/wynntils/features/inventory/ContainerSearchFeature.java:
```diff
@@ -215,91 +235,114 @@ private void tryAutoSearch(AbstractContainerScreen<?> abstractContainerScreen) {
-            if (type.isScreen(screen)) {
-                containerType = type;
+    private SearchableContainerProperty getCurrentSearchableContainer() {
+        // FIXME: Do something about this
```
I guess you could put these in a `Map<Class<? extends SearchableContainerProperty>, Supplier<Boolean>>` but that wouldn't look much nicer, we would just move the code.

ShadowCat117 replied:
I think this looks a lot better than all the if statements

kristofbolyai replied:
Indeed it looks better, all is well then.

kristofbolyai commented on common/src/main/java/com/wynntils/models/containers/BankModel.java:
```diff
@@ -68,24 +69,20 @@ public class BankModel extends Model {
     private boolean editingName;
     private int currentPage = 1;
-    private InteractiveContainerType currentContainer;
+    private PersonalStorageType currentContainer = null;
```
Maybe this field and the getter requires a rename? `storageContainerType`?

kristofbolyai commented on common/src/main/java/com/wynntils/models/containers/ContainerModel.java:
```diff
@@ -64,11 +108,42 @@ public boolean isLootOrRewardChest(String title) {
+    }
+
+    private void registerWynncraftContainers() {
+        // Order does not matter here so just keep it alphabetical
+        registerWynncraftContainer(new AbilityTreeContainer());
+        registerWynncraftContainer(new AccountBankContainer());
+        registerWynncraftContainer(new BlockBankContainer());
+        registerWynncraftContainer(new BookshelfContainer());
+        registerWynncraftContainer(new CharacterBankContainer());
+        registerWynncraftContainer(new ContentBookContainer());
+        registerWynncraftContainer(new GuildBankContainer());
+        registerWynncraftContainer(new GuildMemberListContainer());
+        registerWynncraftContainer(new GuildTerritoriesContainer());
+        registerWynncraftContainer(new HousingJukeboxContainer());
+        registerWynncraftContainer(new HousingListContainer());
+        registerWynncraftContainer(new JukeboxContainer());
+        registerWynncraftContainer(new LobbyContainer());
+        registerWynncraftContainer(new MiscBucketContainer());
+        registerWynncraftContainer(new PetMenuContainer());
+        registerWynncraftContainer(new ScrapMenuContainer());
+        registerWynncraftContainer(new TradeMarketFiltersContainer());
+        registerWynncraftContainer(new TradeMarketPrimaryContainer());
+        registerWynncraftContainer(new TradeMarketSecondaryContainer());
```
Maybe we could keep it alphabetical by type, kind of like features. But we can keep it as is, really doesn't matter.

kristofbolyai commented on common/src/main/java/com/wynntils/models/containers/type/SearchableContainerProperty.java:
```diff
@@ -0,0 +1,11 @@
+public interface SearchableContainerProperty extends ScrollableContainerProperty {
+    ContainerBounds getBounds();
+
+    boolean supportsAdvancedSearch();
```
Did we ever have a way to disable the advanced search with a config? If yes, do you know if that works? (If we have no such config, ignore this :)).

ShadowCat117 replied:
I don't think so, when the advanced search was added it just replaced the original basic search

kristofbolyai commented on common/src/main/java/com/wynntils/models/containers/type/wynncontainers/AbilityTreeContainer.java:
```diff
@@ -0,0 +1,38 @@
+    private static final Pattern PREVIOUS_PAGE_PATTERN = Pattern.compile("§7Previous Page");
+
+    public AbilityTreeContainer() {
+        super(Pattern.compile("(?:Warrior|Shaman|Mage|Assassin|Archer) Abilities"));
```
Although this specific case may not have a test at the moment, can we change the constructor pattern to extract the pattern into a static final field? Otherwise, these regexes become un-testable by our framework. Please change every container class.

kristofbolyai commented on common/src/main/java/com/wynntils/screens/base/widgets/ItemSearchButton.java:
```diff
@@ -0,0 +1,30 @@
+import net.minecraft.client.gui.screens.Screen;
+import net.minecraft.network.chat.Component;
+
+public class ItemSearchButton extends WynntilsButton {
```
`ItemSearchButton` is a bit too broad, I had to read the code to know what this does. I propose renaming to `ItemSearchUIButton`/`ItemFilterUIButton`.

kristofbolyai commented on common/src/main/java/com/wynntils/screens/itemfilter/ItemFilterScreen.java:
```diff
@@ -0,0 +1,1279 @@
+        usedProviders.add(newFilter.a());
+
+        createFilterQuery();
+        populateQuery();
```
Do you not need to call `populateSorts` here and in `removeFilter`? If not, maybe you could include a comment about them below the two calls.

ShadowCat117 replied:
No, because those 2 methods can only be called from the value widgets so you won't be viewing the sorts at the moment

kristofbolyai replied:
If you could add a comment in the code, we can mark this as completed.

kristofbolyai commented on common/src/main/java/com/wynntils/screens/itemfilter/ItemFilterScreen.java:
```diff
@@ -0,0 +1,1279 @@
+        }
+    }
+
+    private void parseFilters(String input) {
```
I would assume this is the hacky part you mentioned before. At first sight, my first idea/question is why not just use `ItemFilterService#createSearchQuery`? You get back a parsed object, from that, you can easily get all providers/sorts/whatever you need. Would this work?

kristofbolyai replied:
I overall think  that the widget's `updateValues` do too much parsing, I assume you know this as well. I wonder if you would still need this parsing if'd done the suggestion above, since `ItemSearchQuery` has the StatFilters too, perhaps it could be passed to the widget to make parsing easier since you would only have to decide how you represent the filter in the widget itself, without needing to parse it.

ShadowCat117 replied:
I think `parseFilters` is fine, it's the `updateValues` I'm not a fan of especially on the `IntegerValueWidget`. I'll look into using the services method, maybe that will be better

ShadowCat117 replied:
I tried using this but was unable to get it to work for filters. Sorts and plain queries were fine but after removing the parsing from the constructors and always calling `updateValues` I'm happier with how it works. The `ignoreUpdate` field in the widgets still bothers me a bit but I don't see how it can be avoided, unless I add a `setTextBoxInputWithoutConsuming`..

kristofbolyai replied:
I'll try to make a patch to make this work.

kristofbolyai commented on common/src/main/java/com/wynntils/screens/itemfilter/widgets/IntegerValueWidget.java:
```diff
@@ -0,0 +1,508 @@
+                lessThanEqual = lessThanMatcher.group(1).equals("<=");
+                this.lessThanPercentage = value.endsWith("%") && supportsPercentage;
+            }
+        }
```
Can't you call `updateValues`?

kristofbolyai commented on common/src/main/java/com/wynntils/services/itemfilter/statproviders/GearTypeStatProvider.java:
```diff
@@ -16,4 +18,23 @@ public List<String> getValue(WynnItem wynnItem) {
+
+    @Override
+    public List<String> getValidInputs() {
+        // Can't use all values of the enum as WEAPON, ACCESSORY, MASTERY_TOME and CHARM
+        // don't work with this filter, ItemTypeStatProvider should be used for tomes and charms
```
`ItemTypeStatProvider` doesn't work for WEAPON, ACCESSORY?

ShadowCat117 replied:
No, you can only do `itemType:gear`, there is no `WeaponItem` or `AccessoryItem`

kristofbolyai replied:
Maybe we should implement that. Probably not here though.

kristofbolyai commented on common/src/main/java/com/wynntils/services/itemfilter/statproviders/ItemTypeStatProvider.java:
```diff
@@ -13,4 +34,32 @@ public class ItemTypeStatProvider extends ItemStatProvider<String> {
+                RuneItem.class.getSimpleName().replace("Item", ""),
+                SimulatorItem.class.getSimpleName().replace("Item", ""),
+                TeleportScrollItem.class.getSimpleName().replace("Item", ""),
+                TomeItem.class.getSimpleName().replace("Item", ""));
```
We will forgot to update this. Can we get all inheritors of WynnItem and display all, even if not all will be "real" items?

ShadowCat117 replied:
I was going to do that originally but I couldn't find a good way to do it

magicus replied:
That seems to be tricky, https://stackoverflow.com/questions/205573/at-runtime-find-all-classes-in-a-java-application-that-extend-a-base-class. 

Since we own the code, can't we change the WynnItem supertype to register themselves automatically?

kristofbolyai replied:
I suggest getting `ItemHandler`'s `ItemAnnotator`s, they have the same name as the items, you just replace the "Annotator" part. Otherwise, they should exist whenever an item exists. You can probably Make two subclasses for `ItemAnnotator`, `GameItemAnnotator`, `GuiItemAnnotator` (and leave the last 3 edge cases as normal implementations).

kristofbolyai commented on common/src/main/resources/assets/wynntils/lang/en_us.json:
```diff
@@ -2166,28 +2199,49 @@
   "service.wynntils.itemFilter.sortingNotSupported": "Sorting is not supported for this container",
   "service.wynntils.itemFilter.stat.actual.description": "%s stat of the item",
   "service.wynntils.itemFilter.stat.countedItem.description": "Count of an item (e.g. health potion), this is not the same as the amount of items in a stack.",
+  "service.wynntils.itemFilter.stat.countedItem.name": "Item Count",
```
I see you added these name i18ns. Did you add an assert for them in the code?

ShadowCat117 replied:
No, the providers for skills and actual stats don't use i18n

kristofbolyai replied:
Oh, okay.

ShadowCat117 committed:
Merge remote-tracking branch 'upstream/main' into filter-gui

ShadowCat117 committed:
chore: Update variable name

ShadowCat117 committed:
chore: Rename BankModel variable and method

ShadowCat117 committed:
chore: Rename ItemSearchButton to ItemFilterUIButton

ShadowCat117 committed:
refactor: Extract title pattern for WynncraftContainers

ShadowCat117 committed:
refactor: Call updateValues in ValueWidget constructors

ShadowCat117 committed:
chore: Add comment about not requiring populateSorts method

ShadowCat117 committed:
feat: Support OR filters

magicus commented:
I'd like to echo bolyai: This looks fantastic, judging from the video! But it is quite a bit to review. I'll try to get to it, but I can't promise when.

ShadowCat117 commented:
> I'd like to echo bolyai: This looks fantastic, judging from the video! But it is quite a bit to review. I'll try to get to it, but I can't promise when.

It's mostly the first commit I'd like your opinion on so if you only have time for that don't worry about the screen itself :)

kristofbolyai commented:
Yes magicus, if you review the screen refactors, I can review the widgets and UI, if you don't have time for it.

ShadowCat117 committed:
refactor: Use supplier map for looking up searchable container

ShadowCat117 committed:
fix: Fix bank features still be active if opening emerald pouch/potion menu from bank

ShadowCat117 committed:
refactor: Sort ItemAnnotator into Game and Gui annotators. Use in ItemTypeStatProvider

kristofbolyai committed:
refactor: Introduce `SortInfo`, use it everywhere applicable

kristofbolyai committed:
feat: Finish refactoring of ItemFilterScreen with object-based filter/sort storage

kristofbolyai committed:
feat: Implement a method for `ItemFilterService` to produce a query string of a parsed query

kristofbolyai committed:
refactor: Add some generics constraints to NumericValueWidget

kristofbolyai committed:
fix: Fix ItemFilterService#getItemFilterString adding leading/trailing spaces unnecessarily

kristofbolyai committed:
fix: Fix NumericValueWidget not adding "*" as the input text

kristofbolyai committed:
fix: Fix NumericValueWidget infinite-recursion with greaterThan/lessThan

kristofbolyai committed:
refactor: Remove separate `usedProviders` set

kristofbolyai committed:
fix: Fix equality buttons not working for NumericValueWidget, fix not resetting input fields

kristofbolyai committed:
refactor: Add an apply button to the main item search widget, when the filter UI is open

kristofbolyai committed:
fix: Fix "any" value not working when only the input is "*" in NumericValueWidget

kristofbolyai committed:
fix: Apply search the first time the UI is opened

kristofbolyai committed:
fix: Fix StatProviderFilterMap using `Collections#unmodifiable*` and changing on us during recursing methods

kristofbolyai committed:
fix: Fix WynntilsCheckbox behavior, fixing value widgets with checkboxes

kristofbolyai committed:
fix: Use short float form, if possible

kristofbolyai committed:
fix: Fix any checkbox not removing "*"

kristofbolyai committed:
fix: Fix ListValueWidget not getting initialized properly

kristofbolyai committed:
fix: Fix StringValueWidget crashing when there are no filters

kristofbolyai committed:
fix: Fix strict mode not working in StringValueWidget

kristofbolyai committed:
fix: Register TargetStatProvider

kristofbolyai committed:
fix: Fix applyButton being active when there are no changes to apply in ItemFilterScreen

kristofbolyai committed:
refactor: Remove unused constructor parameters, make fields final, use correct widget name

kristofbolyai committed:
fix: Fix StringUtils#floatToSimpleString producing weird output due to float error

kristofbolyai committed:
fix: Fix BooleanValueWidget behaviour

kristofbolyai committed:
fix: Fix right click not removing a filter

kristofbolyai committed:
fix: Fix CappedValue widget crashing

ShadowCat117 committed:
fix: Fix percentage greater/less than

ShadowCat117 committed:
Merge pull request #1 from Wynntils/filter-gui-refactor-parsing

refactor: Refactor how the Filter UI handles sort/filter values

Commit:
ci: spotless formatting

ShadowCat117 committed:
feat: Enter to apply search widget changes

ShadowCat117 committed:
fix: Keep apply button active on reinit

ShadowCat117 committed:
feat: Only show sort toggle button when sorting supported, make clearer when in sort/filter mode

ShadowCat117 committed:
fix: Don't add empty providers to search query

ShadowCat117 committed:
feat: Set any filter via middle click

ShadowCat117 committed:
chore: Change middle click tooltip

ShadowCat117 committed:
wip: Support multiple filters for string providers

kristofbolyai commented on common/src/main/java/com/wynntils/screens/itemfilter/widgets/StringValueWidget.java:
```diff
@@ -4,138 +4,439 @@
+                widgets.add(new StringWidget(getX(), renderY, "", false));
+            } else {
+                String currentInput = filters.get(currentValue).statFilter().asString();
+                boolean strict = currentInput.startsWith("\"") && currentInput.endsWith("\"");
```
You could check for `instanceof StringStatFilter`, then call `isStrict`.

kristofbolyai commented on common/src/main/java/com/wynntils/screens/itemfilter/ItemFilterScreen.java:
```diff
@@ -670,6 +670,10 @@ public void setFiltersForProvider(ItemStatProvider<?> provider, List<StatProvide
+
+        if (provider == selectedProvider) {
+            valueWidget.onFiltersChanged(filterPairs);
+        }
```
I can't seem to understand why this is needed. It's also the reason as to why `BooleanValueWidget` needs the null check for filters.

ShadowCat117 replied:
It's so when you interact with the provider button to add the any filter or remove all filters it updates the widget

ShadowCat117 replied:
This has been removed now

ShadowCat117 committed:
fix: Fix crash when clearing string filter whilst selected

ShadowCat117 committed:
refactor: Use instanceof StringStatFilter for strict check

ShadowCat117 committed:
refactor: Call set selected when interacting with provider button

Should function the same as calling onWidgetChanged in setFiltersForProvider but without repopulating the widgets as you interact with them

ShadowCat117 committed:
chore: Remove null checks from boolean and list widgets

ShadowCat117 commented:
Closing for now. Will re open a pr for the `ContainerModel` changes shortly then a new one for the filter GUI.
